tools/python/src/face_detection.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2017  Davis E. King (davis@dlib.net)
+// License: Boost Software License   See LICENSE.txt for the full license.
+
+#include <dlib/python.h>
+#include <boost/shared_ptr.hpp>
+#include <dlib/matrix.h>
+#include <boost/python/slice.hpp>
+#include <dlib/geometry/vector.h>
+#include <dlib/dnn.h>
+#include <dlib/image_transforms.h>
+#include "indexing.h"
+
+using namespace dlib;
+using namespace std;
+using namespace boost::python;
+
+typedef matrix<double,0,1> cv;
+
+
+class face_detection_model_v1
+{
+
+public:
+
+    face_detection_model_v1(const std::string& model_filename)
+    {
+        deserialize(model_filename) >> net;
+    }
+
+    std::vector<rectangle> face_detector (
+        object pyimage,
+        const int upsample_num_times
+    )
+    {
+        // Copy the data into dlib based objects
+        matrix<rgb_pixel> image;
+        if (is_gray_python_image(pyimage))
+            assign_image(image, numpy_gray_image(pyimage));
+        else if (is_rgb_python_image(pyimage))
+            assign_image(image, numpy_rgb_image(pyimage));
+        else
+            throw dlib::error("Unsupported image type, must be 8bit gray or RGB image.");
+
+        // Upsampling the image will allow us to detect smaller faces but will cause the
+        // program to use more RAM and run longer.
+        float scale = 1.0;
+        unsigned int levels = upsample_num_times;
+        while (levels > 0)
+        {
+            levels--;
+            scale *= 2;
+            pyramid_up(image);
+        }
+
+        auto dets = net(image);
+        std::vector<rectangle> rects;
+
+        // Scale the detection locations back to the original image size
+        // if the image was upscaled.
+        for (auto&& d : dets) {
```

Comment:
This should use the built in rect_up() functions.  You make a pyramid_down<2> object and then you can call rect_up() on it and it will do the right thing.  All the pyramid stuff has these things and they take care to deal with any border padding and other vagaries of this coordinate transform.